Output messages grouping

v1nt06 opened the pull request:
Hello Damian,

About 3 months ago I created same pull request, but I was busy to fix changes after your review ([previous pull request with my new replies](https://github.com/damianszczepanik/cucumber-reporting/pull/518#pullrequestreview-3009123)). So, now I complete this and rewrote some code. Could you take a look and accept pull request if everything is good?  Current functional changes in my fork:

- Added grouping for output messages inside steps
- Outputs from different sources grouped separately
- Added word wrapping for messages and errors
- Error message title contains description (1st line of full error message, e.g. exception name)
- Error messages and outputs is expanded by default

Sample.json with this functional changes:
![sample json](https://cloud.githubusercontent.com/assets/2842503/22290622/5974337c-e31b-11e6-875f-be29fc26ef84.png)

Test scenario with different outputs:
![expample](https://cloud.githubusercontent.com/assets/2842503/22290631/68aad1a2-e31b-11e6-9df1-4149733f7390.png)

damianszczepanik commented on src/main/java/net/masterthought/cucumber/json/deserializers/StepDeserializer.java:
```diff
@@ -0,0 +1,65 @@
+public class StepDeserializer extends JsonDeserializer<Step> {
+
+	@Override
+	public Step deserialize(JsonParser parser, DeserializationContext context) throws IOException {
```
instead of providing deserializer that makes code ugly, maybe try to provide additional method such as getOutputs() that wraps it? Then the model stays the  same as iti s in json file and onlyl view is using method for displaying content in different way

v1nt06 replied:
We can't wraps it because in this case we can't determine what was in json. For example, we can't determine what was in json: two diffirent outputs with one message in each or one output with two messages. So we need to detect this situations in deserialization process.

damianszczepanik commented on src/main/resources/templates/macros/json/output.vm:
```diff
@@ -1,16 +1,21 @@
-          <a>Output $index</a>
+        <div data-toggle="collapse" class="#if ($isPassed) collapsed #end collapsable-control" data-target="#output-$outputId">
+          Output $index
+          <i class="chevron fa fa-fw"></i>
```
what is it used for?

v1nt06 replied:
For collapsing functionality for outputs at the same style like steps. Please see screenshots above.

damianszczepanik commented on src/main/resources/templates/macros/json/output.vm:
```diff
@@ -1,16 +1,21 @@
+#if($outputs)
   <div class="outputs inner-level">
-    #foreach($message in $output.getMessages())
+    #foreach($output in $outputs)
```
provided screenshtos don't really shows what is your intention. Can you describe it?

v1nt06 replied:
Sure. Here is demo screenshots.
Current output:
![old](https://cloud.githubusercontent.com/assets/2842503/22456431/d0613fa4-e7ab-11e6-964b-dc7e21cf3f71.png)

New output:
![new](https://cloud.githubusercontent.com/assets/2842503/22456438/da019d88-e7ab-11e6-96d1-8214ccb71143.png)

damianszczepanik replied:
Looking at this I wonder why not having single output section?

damianszczepanik commented on src/test/java/net/masterthought/cucumber/json/StepTest.java:
```diff
@@ -65,10 +65,10 @@ public void getOutput_ReturnsOutput() {
 
         // then
-        assertThat(output.getMessages()).containsOnlyOnce(
+        assertThat(getMessages(outputs)).containsOnlyOnce(
```
finally all commits should be squals into one

v1nt06 replied:
Ok.

damianszczepanik replied:
generally I'm OK with this feature as long as this does not generate more code that is needed...

v1nt06 commented:
All changes in one commit was added above. (4c035a9)

damianszczepanik commented on src/main/java/net/masterthought/cucumber/json/Step.java:
```diff
@@ -1,22 +1,33 @@
+    private final Row[] rows;
+    private final Match match;
+    private final Embedding[] embeddings;
+    private final Output[] outputs;
```
have you tried to serialize using Object type? then the rest is used as now and the only implementation you need to provide is for single or array outputs

AFAIR this is how it was deserialized at the very beggining

v1nt06 replied:
I returned previous implementation and created deserializer only for output node.

damianszczepanik commented on src/main/java/net/masterthought/cucumber/json/deserializers/StepDeserializer.java:
```diff
@@ -0,0 +1,65 @@
+        Output[] outputs = parseOutputs(rootNode.get("output"));
+
+        return new Step(name, keyword, result, rows, match, embeddings, docString, outputs);
+    }
```
I want to minimalize code and execution that can be performed by 3td party lib

v1nt06 replied:
No longer relevant. This code was deleted after creating deserializer only for output node.

damianszczepanik commented on src/test/java/net/masterthought/cucumber/generators/integrations/PageTest.java:
```diff
@@ -38,4 +42,13 @@ protected DocumentAssertion documentFrom(String pageName) {
+    protected String[] getMessages(Output[] outputs) {
+        List<String> messages = new ArrayList<>();
+        for (Output output : outputs) {
+            messages.addAll(Arrays.asList(output.getMessages()));
```
I don't see unit test for this code

v1nt06 replied:
This method used only in test and designed for getting output messages for step tests and feature report page integration tests.

v1nt06 commented:
Damian, I've pushed new changes. Could you take a look please?

damianszczepanik commented on src/main/java/net/masterthought/cucumber/json/Result.java:
```diff
@@ -28,4 +28,8 @@ public String getFormatedDuration() {
     }
+
+    public final String getErrorMessageTitle() {
+        return errorMessage.split("[\\r\\n]+")[0];
```
Based on https://docs.oracle.com/javase/7/docs/api/java/util/regex/Pattern.html
\s should be enough

v1nt06 replied:
If we put into pattern "\s" we got only first word from message. Current pattern returns first line. For example, we have exception. In case with "\s" pattern we got only type of exception, but with current pattern we got type with message in one line. I think it is more useful.

damianszczepanik commented on src/main/java/net/masterthought/cucumber/json/Step.java:
```diff
@@ -13,7 +15,9 @@
-    private final Output output = null;
+    @JsonDeserialize(using = OutputsDeserializer.class)
+    @JsonProperty("output")
+    private final Output[] outputs = new Output[0];
```
Have you tried to use single class and deserialilze it inside? so line 19 won't be required and deserializer method will be linked iwth Outputs class. Then also there will be single attribute here Objects, not the array

So it is more like the current solution

v1nt06 replied:
No, I haven't tried. Because in this case we need to collect outputs (if we have more than 1 source) inside Output class in array and every element of this array should contain messages. But I can do this. What I should do?

damianszczepanik commented on src/main/resources/templates/macros/json/output.vm:
```diff
@@ -1,16 +1,21 @@
-    #foreach($message in $output.getMessages())
+    #foreach($output in $outputs)
+      #set($outputId = $counter.next())
+      #set($index = $foreach.index + 1)
```
why not using counter as previously?

v1nt06 replied:
Ok, I'll use only one counter.

v1nt06 replied:
> Looking at this I wonder why not having single output section?
Because we can't determine from wich output we got messages if we have few output sources. Do you mean that we need to group two different output sources into one like this? (two options):

1. 
![1](https://cloud.githubusercontent.com/assets/2842503/22972264/5553e6a2-f392-11e6-806d-9741e2ff5904.png)
2. 
![2](https://cloud.githubusercontent.com/assets/2842503/22972268/5a86cc48-f392-11e6-9cc2-0158880db035.png)


> why not using counter as previously?
Because "index"-counter count output number in every step. "outputId"-counter need for collapsible sections.

damianszczepanik replied:
Please try

damianszczepanik commented on src/test/java/net/masterthought/cucumber/generators/AbstractPageTest.java:
```diff
@@ -66,7 +66,7 @@ public void generateReport_DisplaysContentAsEscapedText() {
                 .isEqualTo("Attachment 4 (HTML) <i>Hello</i> <b>World!</b>");
         assertThat(document.getFeature().getElements()[0].getStepsSection().getSteps()[5].getMessage().text())
-                .isEqualTo("Error message java.lang.AssertionError: \n" +
+                .isEqualTo("java.lang.AssertionError: java.lang.AssertionError: \n" +
```
if the input data is not changed, why expected is?

v1nt06 replied:
Input data was chnged in 14th line in steps.vm and in getErrorMessageTitle() inside Result class.

damianszczepanik commented:
Rebase the code, squash commits and let me know how about this deserializer.

v1nt06 commented:
Rebased and squashed. Deserializer remains as is. I wrote reason above:
https://github.com/damianszczepanik/cucumber-reporting/pull/594#discussion_r101259353

damianszczepanik commented:
I don't have more comments, thanks!
